ldap/servers/slapd/back-ldbm/ldbm_search.c

```diff
@@ -1695,31 +1694,41 @@ ldbm_back_next_search_entry(Slapi_PBlock *pb)
                 filter_test = -1;
             } else {
                 /* it's a regular entry, check if it matches the filter, and passes the ACL check */
-                if (0 != (sr->sr_flags & SR_FLAG_CAN_SKIP_FILTER_TEST)) {
-                    /* Since we do access control checking in the filter test (?Why?) we need to check access now */
+                if (0 == (sr->sr_flags & SR_FLAG_MUST_APPLY_FILTER_TEST)) {
+                    /* Since we do access control checking in the filter test we need to check access now */
                     slapi_log_err(SLAPI_LOG_FILTER, "ldbm_back_next_search_entry",
                                   "Bypassing filter test\n");
                     if (ACL_CHECK_FLAG) {
-                        filter_test = slapi_vattr_filter_test_ext(pb, e->ep_entry, filter, ACL_CHECK_FLAG, 1 /* Only perform access checking, thank you */);
+                        filter_test = slapi_vattr_filter_test_ext(pb, e->ep_entry, filter_intent, ACL_CHECK_FLAG, 1 /* Only perform access checking, thank you */);
```

Comment:
I am missing something, filter_intent is the original and optimized filter but it does not contain (objectclass=referral). If the candidate is a referral, this filter will not match and the referral will be skipped. ?

Reply:
we skip checking it against *aci* specifically. because the user may not have the aci to read referals. The same way the user may not have the aci to see parentid. But we still need to enforce this condition in execution ....